### PR TITLE
TEE supplicant support

### PR DESCRIPTION
**Pull request by Deedone**

### CODE REVIEW

**Comment by oleksiimoisieiev** [comment]
Compliance check was failed

File: drivers/tee/optee/optee.c [hunk 1]
```diff
@@ -59,11 +59,34 @@ struct optee_notify {
 	struct k_sem wait;
 };
 
+struct optee_supp_req {
+	sys_dnode_t link;
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c:63]: s/syscals/syscalls/g in commit description

File: drivers/tee/optee/optee.c [hunk 2]
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
 
 static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf);
 
+uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t num_params,
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: static maybe? Is it used from outside?

File: drivers/tee/optee/optee.c [hunk 3]
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
...
+			struct tee_param *param)
+
+{
+	struct optee_driver_data *data = (struct optee_driver_data*) dev->data;
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: should be space between ) and dev->data? I'm not sure, please consult with checkpatch

File: drivers/tee/optee/optee.c [hunk 4]
```diff
@@ -449,6 +511,31 @@ static void free_shm_pages(void **pages)
 	}
 }
 
+static void handle_rpc_supp_cmd(const struct device *dev, struct optee_msg_arg *arg)
+{
+	// TODO: FIX HARDCODE
+	struct tee_param params[5];
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: Hm. It's not clear to me why have you made this hardcode? IIUC we have num_params passed in arg argument.
At least if there is a good reason why have you hardcoded this param I'd like to see it in TODO.

File: drivers/tee/optee/optee.c [hunk 5]
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+			p->b = param[n].b;
+			p->c = param[n].c;
+			break;
+		default:
+			break;
+		}
+	}
+	req->ret = ret;
+
+	supp->current = NULL;
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: This probably should be under the mutex as well. Maybe it's better to make supp->current= NULL right after reading it's value so revc won't get BUSY?

File: tests/drivers/tee/optee/src/main.c [hunk 6]
```diff
@@ -470,6 +470,149 @@ ZTEST(optee_test_suite, test_func_shm_alloc)
...
+	const struct device *const dev = DEVICE_DT_GET_ONE(linaro_optee_tz);
+	struct tee_param params[TEE_NUM_PARAMS];
+	unsigned int num_params = TEE_NUM_PARAMS;
+	uint32_t func;
+	int ret;
+	supp_thread_ok = true;
+
+	ret = tee_suppl_recv(dev, &func, &num_params, params);
+
+	if (ret != 0) {
```
Comment (oleksiimoisieiev) [inline tests/drivers/tee/optee/src/main.c:551]: why not using z_assert here to check return values?

File: drivers/tee/optee/optee.c [hunk 7]
```diff
@@ -282,19 +316,17 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
 		return;
 	}
 
-	/*
-	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
-	 * Currently TYPE_KERNEL was implemented.
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
+		/* todo handle situation when shm was allocated statically so buffer can be reused*/
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: s/todo/TODO/g

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: drivers/tee/optee/optee.c [hunk 8]
```diff
@@ -59,11 +59,34 @@ struct optee_notify {
...
+struct optee_supp_req {
+	sys_dnode_t link;
+
+	bool in_queue;
+	uint32_t func;
+	uint32_t ret;
+	size_t num_params;
+	struct tee_param *param;
+
+	struct k_sem c;
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: Strange naming, purpose is not clear.

File: drivers/tee/optee/optee.c [hunk 9]
```diff
@@ -156,7 +179,7 @@ static int msg_param_to_param(struct tee_param *param, unsigned int num_param,
 		case OPTEE_MSG_ATTR_TYPE_VALUE_INPUT:
 		case OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT:
 		case OPTEE_MSG_ATTR_TYPE_VALUE_INOUT:
-			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT + attr -
+			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INPUT + attr -
```
Comment (firscity) [inline drivers/tee/optee/optee.c:182]: It should be mentioned in commit message or be a separate commit

File: drivers/tee/optee/optee.c [hunk 10]
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
...
+			struct tee_param *param)
+
+{
+	struct optee_driver_data *data = (struct optee_driver_data *) dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req;
+	uint32_t ret;
+
+	req = k_malloc(sizeof(*req));
+	if (!req)
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: Add braces {} for all one-line blocks (checpatch.pl should be used with `--ignore BRACES`)

File: drivers/tee/optee/optee.c [hunk 11]
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+	struct optee_supp_req *req = NULL;
+
+	if (supp->current) {
+		return -EBUSY;
+	}
+
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		if (!sys_dlist_is_empty(&supp->reqs)) {
+			req = (struct optee_supp_req *)sys_dlist_get(&supp->reqs);
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: redundant empty line

File: drivers/tee/optee/optee.c [hunk 12]
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	supp->current = req;
+	k_mutex_unlock(&supp->mutex);
+
+	*func = req->func;
+	*num_params = req->num_params;
+	memcpy(param, req->param,
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: may be a single line

File: drivers/tee/optee/optee.c [hunk 13]
```diff
@@ -282,19 +316,17 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
 		return;
 	}
 
-	/*
-	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
-	 * Currently TYPE_KERNEL was implemented.
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: Maybe switch/case?

File: drivers/tee/optee/optee.c [hunk 14]
```diff
@@ -323,27 +355,27 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
...
 		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
 		return;
 	}
 
-	/*
-	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
-	 * Currectly TYPE_KERNEL was implemented.
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: same about switch/case

File: tests/drivers/tee/optee/src/main.c (hunk 6 above)
Comment (Deedone) [reply tests/drivers/tee/optee/src/main.c:551]: It causes the main thread to hang when assert fails.

Comment (oleksiimoisieiev) [reply tests/drivers/tee/optee/src/main.c:551]: +1 let's leave as is

File: drivers/tee/optee/optee.c (hunk 9 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c:182]: Moved to a separate commit

File: drivers/tee/optee/optee.c [hunk 15]
```diff
@@ -476,8 +571,8 @@ static uint32_t handle_func_rpc_call(const struct device *dev, struct tee_shm *s
 		return TEEC_ERROR_NOT_IMPLEMENTED;
 	default:
 		/* TODO: supplicant commands are still not implemented */
```
Comment (oleksiimoisieiev) [inline drivers/tee/optee/optee.c]: Probably no use for this todo because commands were implemented.

**Comment by oleksiimoisieiev** [comment]
LGTM after removing redundand TODO
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

File: drivers/tee/optee/optee.c [hunk 16]
```diff
@@ -449,6 +512,38 @@ static void free_shm_pages(void **pages)
...
+		goto out;
+	}
+
+	arg->ret = optee_call_supp(dev, arg->cmd, arg->num_params, params);
+
+	ret = param_to_msg_param(params, arg->num_params, arg->params);
+	if (ret) {
+		arg->ret = TEEC_ERROR_GENERIC;
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: nit: no need for this goto

File: drivers/tee/optee/optee.c [hunk 17]
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
...
+	struct tee_param param;
+
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
+	param.a = OPTEE_RPC_SHM_TYPE_APPL;
+	param.b = sz;
+	param.c = 0;
+
+	ret = optee_call_supp(dev, OPTEE_RPC_CMD_SHM_ALLOC, 1, &param);
+
+	if (ret)
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: braces

File: drivers/tee/optee/optee.c [hunk 18]
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
...
+	ret = tee_add_shm(dev, (void *)param.c, 0, param.b, 0, shm);
+
+	return ret;
+}
+
+static void cmd_free_suppl(const struct device *dev, struct tee_shm *shm)
+{
+	struct tee_param param;
+
+
```
Comment (firscity) [inline drivers/tee/optee/optee.c]: empty line

File: drivers/tee/optee/optee.c [hunk 19]
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
...
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
```
Comment (firscity) [inline drivers/tee/optee/optee.c:297]: may be done in initialization:
`... param = {
    .attr = ...
};`

File: tests/drivers/tee/optee/src/main.c [hunk 20]
```diff
@@ -617,6 +617,180 @@ ZTEST(optee_test_suite, test_func_rpc_supp_cmd)
...
+
+	if (params[0].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INOUT ||
+	    params[0].a != OPTEE_RPC_SHM_TYPE_APPL) {
+		printk("Unexpected params %d %d\n", params[0].attr,
+		       params[0].a);
+		supp_thread_ok = false;
+		goto fail1;
+	}
+
+	mem = k_malloc(params[0].b);
```
Comment (firscity) [inline tests/drivers/tee/optee/src/main.c:725]: check return?

File: drivers/tee/optee/optee.c (hunk 19 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c:297]: I think it looks better with separate declaration and initialization.

File: drivers/tee/optee/optee.c (hunk 16 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c]: Let's leave it in case someone adds some new code.

File: drivers/tee/optee/optee.c [hunk 21]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
-static int optee_suppl_send(const struct device *dev, unsigned int num_params,
+static int optee_suppl_send(const struct device *dev, unsigned int ret, unsigned int num_params,
 			    struct tee_param *param)
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req;
+	size_t n;
+
+	if (!supp->current)
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: I believe that we can drop this check

File: drivers/tee/optee/optee.c [hunk 22]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+
+	*func = req->func;
+	*num_params = req->num_params;
+	memcpy(param, req->param, sizeof(struct tee_param) * req->num_params);
+
 	return 0;
 }
 
-static int optee_suppl_send(const struct device *dev, unsigned int num_params,
+static int optee_suppl_send(const struct device *dev, unsigned int ret, unsigned int num_params,
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c:1059]: do we really need `num_params`? should you check that we have enough number of items to hold `req->num_params`
as far as I see, it is unused

File: drivers/tee/optee/optee.c [hunk 23]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+
+		k_sem_take(&supp->reqs_c, K_FOREVER);
+	}
+
+	if (*num_params < req->num_params) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	supp->current = req;
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: can be a situation when we still have another valid/non-handled  `current`?
could you please add assert here at least

File: drivers/tee/optee/optee.c (hunk 16 above)
Comment (xakep-amatop) [reply drivers/tee/optee/optee.c]: someone can add `goto` if he or she add a new chunk of code

File: drivers/tee/optee/optee.c [hunk 24]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
 
-static int optee_suppl_recv(const struct device *dev, uint32_t func, unsigned int num_params,
+static int optee_suppl_recv(const struct device *dev, uint32_t *func, unsigned int *num_params,
 			    struct tee_param *param)
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	if (supp->current) {
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: looks like, it isn't safe
what if somebody sets `supp->current` right after this check from another thread?

File: drivers/tee/optee/optee.c [hunk 25]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+
+		if (req) {
+			break;
+		}
+
+		k_sem_take(&supp->reqs_c, K_FOREVER);
+	}
+
+	if (*num_params < req->num_params) {
+		return -EINVAL;
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: should we process all list items in order or if we have enough space in `param` array for another item from list we can process it right now ? 
in any case, could you please add a few words about it to the code ?

Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: should you put back to the list `req` or, at least, notify the waiting thread about completion??

File: drivers/tee/optee/optee.c [hunk 26]
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+		k_mutex_unlock(&supp->mutex);
+
+		if (req) {
+			break;
+		}
+
+		k_sem_take(&supp->reqs_c, K_FOREVER);
+	}
+
+	if (*num_params < req->num_params) {
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: isn't better to move this check and assignment to the above loop?
I mean is it good idea to break mutex lock? why can it be done under one lock in one place?

**Review by xakep-amatop — COMMENTED** [review]
the review hasn't been finished yet
I've posted some questions just to speed up addressing of issues if there are any

**Comment by xakep-amatop** [comment]
only for two next commits:  538c0381ba385fe51eb9b0aab92889a57ca66f02 and b6485303df678d5e5c1594be61e7b65c05371c5f
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`

File: drivers/tee/optee/optee.c (hunk 22 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c:1059]: Added check

File: drivers/tee/optee/optee.c [hunk 27]
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+	supp->current = NULL;
+	k_mutex_unlock(&supp->mutex);
+
+	if (!req) {
+		return -EINVAL;
+	}
+
+	if (num_params != req->num_params) {
+		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
+			num_params);
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: should you add here the next chunk of code:
```
req->ret = -EINVAL;
k_sem_give(&req->complete);
```
??

I think you should to wake up the waiting thread, but I'm not sure about the error code setting like above
or caller should make this call again with bigger params array

**Review by firscity — APPROVED** [review]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

File: drivers/tee/optee/optee.c [hunk 28]
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		req = (struct optee_supp_req *)sys_dlist_peek_head(&supp->reqs);
+
+		if (req && supp->current == NULL) {
+			if (*num_params < req->num_params) {
+				LOG_ERR("Not enough space for params, need at least %lu",
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c:1034]: it is up to you, because I don't know what will be better
but I think in this case caller should alloc more params and call this function again or you can drop entry from the list and notify the waiting thread

File: drivers/tee/optee/optee.c [hunk 29]
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		req = (struct optee_supp_req *)sys_dlist_peek_head(&supp->reqs);
+
+		if (req && supp->current == NULL) {
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]: so, can be a situation when you have `req != NULL` and `supp->current != NULL`
than the below check:
```
if (req)
```
finishes the loop 

is it what we want ?

File: drivers/tee/optee/optee.c (hunk 27 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c]: An invalid number of parameters here indicates that there is a response for some other request, so I think it's better to just ignore it.

File: drivers/tee/optee/optee.c (hunk 28 above)
Comment (Deedone) [reply drivers/tee/optee/optee.c:1034]: With this new logic, the request is left in the queue in case of error, awaiting correct retrieval by the supplicant.

Comment (xakep-amatop) [reply drivers/tee/optee/optee.c:1034]: ok, got it, I hope it doesn't cause the stall

File: drivers/tee/optee/optee.c (hunk 27 above)
Comment (xakep-amatop) [reply drivers/tee/optee/optee.c]: who will notify the waiting thread in that case?

Comment (Deedone) [reply drivers/tee/optee/optee.c]: It will be notified when a response with the correct number of parameters arrives.

Comment (xakep-amatop) [reply drivers/tee/optee/optee.c]: how is it not possible? where can you get ptr to req?
as far as I see, you erased `supp->current` a few lines above

so, req not in queue and not in `supp->current`

Comment (Deedone) [reply drivers/tee/optee/optee.c]: You are right, didn't catch that.

File: drivers/tee/optee/optee.c (hunk 19 above)
Comment (xakep-amatop) [reply drivers/tee/optee/optee.c:297]: note: here can be a leak of some data, also if we have more fields that you initialize here some of fields will contain garbage and it can influence working of OPTEE OS

initializers are like: `{}`, `{0}` and `{.attr = ...}` solve the last issue, but they still have security issues in case when we have paddings between fields in structure, they can contain some stack garbage and it can be useful for some hacker's attacks

so, the best way is to use here `memset` for `param`
but I don't see how somebody can use stack garbage in structure paddings in this case

File: drivers/tee/optee/optee.c [hunk 30]
```diff
@@ -205,6 +228,80 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
...
+static void cmd_free_suppl(const struct device *dev, struct tee_shm *shm)
+{
+	struct tee_param param;
+
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
+	param.a = OPTEE_RPC_SHM_TYPE_APPL;
+	param.b = (uint64_t)shm;
+	param.c = 0;
+
+	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
```
Comment (sa-kib) [inline drivers/tee/optee/optee.c:302]:
```suggestion
	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
	tee_rm_shm(dev, shm);
```

we must free `shm` allocated in `cmd_alloc_suppl()`

File: drivers/tee/optee/optee.c [hunk 31]
```diff
@@ -205,6 +228,46 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
...
+	req->func = func;
+	req->num_params = num_params;
+	req->param = param;
+
+	/* Insert the request in the request list */
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	sys_dlist_append(&supp->reqs, &req->link);
+	k_mutex_unlock(&supp->mutex);
+
+	/* Tell an eventual waiter there's a new request */
```
Comment (lorc) [inline drivers/tee/optee/optee.c]: "event listener"

"Eventual waiter" is something completely different : )

File: include/zephyr/drivers/tee.h [hunk 32]
```diff
@@ -349,7 +349,7 @@ typedef int (*tee_shm_unregister_t)(const struct device *dev, struct tee_shm *sh
  *
  * See @a tee_suppl_recv() for argument definitions.
  */
-typedef int (*tee_suppl_recv_t)(const struct device *dev, uint32_t func, unsigned int num_params,
+typedef int (*tee_suppl_recv_t)(const struct device *dev, uint32_t *func, unsigned int *num_params,
```
Comment (lorc) [inline include/zephyr/drivers/tee.h:352]: Changes in this file should go into a separate commit with a note that they should be squashed into original commit during upstreaming process

File: drivers/tee/optee/optee.c [hunk 33]
```diff
@@ -876,15 +1011,101 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
...
+
+	if (!req) {
+		return -EINVAL;
+	}
+
+	if (num_params != req->num_params) {
+		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
+			num_params);
+		return -EINVAL;
+	}
```
Comment (xakep-amatop) [inline drivers/tee/optee/optee.c]:
```suggestion
	struct optee_supp_req *req = NULL;
	size_t n;

	k_mutex_lock(&supp->mutex, K_FOREVER);
	if (supp->current && num_params >= supp->current->num_params) {
		req = supp->current;
		supp->current = NULL;
	} else {
		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
			num_params);
	}
	k_mutex_unlock(&supp->mutex);

	if (!req) {
		return -EINVAL;
	}
```

**Review by xakep-amatop — APPROVED** [review]
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>` commit 21d89b4a45af7f14e5daf9959aa683a641d5422b
I have finished review

**Review by lorc — APPROVED** [review]
`Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>`
